File: docs/sphinx/examples/python/measuring_kernels.py
# [Begin Docs]
import cudaq
# [End Docs]


# [Begin Sample1]
@cudaq.kernel
def kernel():
    qubits = cudaq.qvector(2)
    mz(qubits)


# [End Sample1]


# [Begin Sample2]
@cudaq.kernel
def kernel():
    qubits_a = cudaq.qvector(2)
    qubit_b = cudaq.qubit()
    mz(qubits_a)
    mx(qubit_b)


# [End Sample2]


# [Begin Run0]
@cudaq.kernel
def kernel() -> list[bool]:
    q = cudaq.qvector(2)

    h(q[0])
    b0 = mz(q[0])
    reset(q[0])
    x(q[0])

    if b0:
        h(q[1])

    return mz(q)


from collections import Counter

results = cudaq.run(kernel, shots_count=1000)
# Convert results to bitstrings and count
bitstring_counts = Counter(
    ''.join('1' if bit else '0' for bit in result) for result in results)

print(f"Bitstring counts: {dict(bitstring_counts)}")

# [End Run0]

''' [Begin Run1]
Bitstring counts: {'11': 247, '10': 753}
 [End Run1] '''

File: docs/sphinx/examples/cpp/measuring_kernels.cpp
/*******************************************************************************
 * Copyright (c) 2025 - 2026 NVIDIA Corporation & Affiliates.                  *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/

// [Begin Docs]
#include <cudaq.h>
// [End Docs]

// [Begin Sample1]
__qpu__ void kernel0() {
  cudaq::qvector qubits(2);
  mz(qubits[0]);
}
// [End Sample1]

// [Begin Sample2]
__qpu__ void kernel1() {
  cudaq::qvector qubits_a(2);
  cudaq::qubit qubits_b;
  mz(qubits_a);
  mx(qubits_b);
}
// [End Sample2]

// [Begin Run0]
__qpu__ auto kernel2() {
  cudaq::qvector q(2);
  h(q[0]);
  auto b0 = mz(q[0]);
  cudaq::reset(q[0]);
  x(q[0]);

  if (b0) {
    h(q[1]);
  }
  return mz(q);
}

int main() {
  auto results = cudaq::run(1000, kernel2);
  // Count occurrences of each bitstring
  std::map<std::string, std::size_t> bitstring_counts;
  for (const auto &result : results) {
    std::string bits = std::to_string(result[0]) + std::to_string(result[1]);
    bitstring_counts[bits]++;
  }

  printf("Bitstring counts:\n{\n");
  for (const auto &[bits, count] : bitstring_counts) {
    printf("  %s: %zu\n", bits.c_str(), count);
  }
  printf("}\n");

  return 0;
}
// [End Run0]

/* [Begin Run1]
Bitstring counts:
{
  10: 771
  11: 229
}
 [End Run1] */
